Kernel/scheduler.c

```diff
@@ -0,0 +1,175 @@
+#include "time.h"
+#include "scheduler.h"
+#include "graphics.h"
+#include "memoryManager.h"
+
+#define MAX_PROCESSES 16
+#define QUANTUM 4
+#define PROCESS_STACK_SIZE 4096  
+
+typedef enum status {READY = 0, BLOCKED, KILLED} TprocessStatus;
+
+typedef struct processContext {
+    uint8_t priority;
+    TprocessStatus status;
+    uint64_t * topRSP;
+    uint64_t * baseRSP;
+    uint64_t * currentRSP;
+    uint64_t functionAddress;
+} TprocessBlock;
+
+static TprocessBlock * processes[MAX_PROCESSES];
+static uint8_t currentRunningProcessPID = 0;
+
+static int validPID(uint8_t pid);
+
+void sch_init(uint64_t functionAddress, uint16_t priority, int argq, char * args[]){
+
+    // First process to run is the shell (SampleCodeModule)
+    int PID = loadProcess(functionAddress, priority, argq, args);
+    if(-1 == PID){
+        return -1;
+    }
+
+    currentRunningProcessPID = PID;
+
+    // ¿Open here process context? (Check Intel TSS)
+
+    return 0;
+}
+
+int sch_loadProcess(uint64_t functionAddress, uint16_t priority, int argq, char * args[]){
+
+    int i;
+    for (; i<MAX_PROCESSES && processes[i] != NULL; i++);
+
+    if (i == MAX_PROCESSES)
+        return -1;
+
+    TprocessBlock * pblock = mm_malloc(sizeof(TprocessBlock));
+
+    if (pblock == NULL)
+        return -1;
+
+    uint64_t topRSP = mm_malloc(PROCESS_STACK_SIZE);
+
+    if (topRSP == NULL) {
+        free(pblock);
+        return -1;
+    }
+
+    // TO DO: Define priority set (0,1,2, ...)
+    pblock->priority = priority;
+    pblock->status = READY;
+    pblock->functionAddress = functionAddress;
+    pblock->topRSP = topRSP;
+    pblock->baseRSP = topRSP + PROCESS_STACK_SIZE - 1;
+    pblock->currentRSP = // TO DO: Build process context (Check Intel TSS)
+
+    if(pblock-> currentRSP == NULL){
+        return -1;
+    }
+
+    processes[i] = pblock;
+    return i;
+}
+
+uint64_t sch_switchProcess(uint64_t * currentRSP){
+    processes[currentRunningProcessPID]->currentRSP = currentRSP;
+    // TO DO
+
+    /* Main idea:
+      Save running process currentStackPointer
+      Check if quantum has finished or there is another process with higher priority to preempt 
+      Check process status
+      If we have to switch a process, we return the currentStackPointer of the other process.
+      If not, we return currentStackPointer.
+    */
+}
+
+uint8_t sch_getPID(){
+    return currentRunningProcessPID;
+}
+
+int sch_kill(uint8_t pid){
+
+    if(!validPID(pid)){
+        return -1;
+   } 
+
+   // What happens if pid = SampleCodeModuleProcess?
+   if(processes[pid]->status == READY){
+        processes[pid]->status = KILLED;
+        return 0;
+   } else {
+        return -1;
+   }
+}
+
+int sch_block(uint8_t pid){
+
+   if(!validPID(pid)){
+        return -1;
+   } 
+   
+   // What happens if pid = SampleCodeModuleProcess?
+   if(processes[pid]->status == READY){
+        processes[pid]->status = BLOCKED;
+        return 0;
+   } else {
+        return -1;
+   }
```

Comment:
If calling some of these functions during an interruption, should we change a process' status to `READY`? We're about to hand control right back to it anyways